pkg/network/usm/shared_libraries.go

```diff
@@ -249,27 +250,34 @@ func (w *soWatcher) Start() {
 			return nil
 		}
 
-		// report silently parsing /proc error as this could happen
-		// just exit processes
-		proc, err := process.NewProcess(int32(pid))
+		mapsPath := fmt.Sprintf("%s/%d/maps", w.procRoot, pid)
+		maps, err := os.Open(mapsPath)
 		if err != nil {
 			log.Debugf("process %d parsing failed %s", pid, err)
 			return nil
 		}
-		mmaps, err := proc.MemoryMaps(true)
-		if err != nil {
-			if log.ShouldLog(seelog.TraceLvl) {
-				log.Tracef("process %d maps parsing failed %s", pid, err)
-			}
-			return nil
-		}
-
-		root := fmt.Sprintf("%s/%d/root", w.procRoot, pid)
-		for _, m := range *mmaps {
-			for _, r := range w.rules {
-				if r.re.MatchString(m.Path) {
-					w.registry.register(root, m.Path, uint32(pid), r)
-					break
+		defer maps.Close()
+
+		scanner := bufio.NewScanner(bufio.NewReader(maps))
+
+		cache := make(map[string]struct{})
+		for scanner.Scan() {
```

Comment:
💬 suggestion, If we using a linux only implementation as we don't relay over gopsutil package
It would be nice to use mmap https://pkg.go.dev/golang.org/x/exp/mmap as reader

Reply:
we are not relying on gopsutil anymore
and I checked the mmap, it returns ReaderAt, while the scanner requires to have a Reader.

Unless you're suggesting to implement the scanning ourselves, I don't see how it can work

Reply:
💬 suggestion, and alternatively we could just regex `/\(.*)(libssl.so|libcrypto.so...)` the whole mmap would be faster than parsing line by line as we only care about the filepath

Reply:
we need to know which regex matched, to apply the correct hooks
we can have multiple interesting .so files loaded into the same binary

generally I don't follow the suggestion and your intention here

Reply:
https://pkg.go.dev/regexp#FindAllString would take care of that for us, w.rules[].re could used to build the regex
The idea here would be to extract the filepath in one pass and would be faster thanks to Boyer-Moore

Reply:
to do so, we need to read the entire maps file to the memory (wasteful)
furthermore, we'll still to iterate the results and to match them to the different rules to know which hooks to apply

Reply:
talked by heart, we don't see the actual benefit and how to use that